### PR TITLE
Min distance filter (and other filter improvments)

### LINKED ISSUE #273: Merge point
Just noticed a bad thing on the Cutter movements.

It does a lot of "pen UP; pen Down" that probably Inkcut should remove.

It starts within Inkscape. You can draw a path with segment, the ending point of one segment  being the start of the next segment.

Almost invisible in Inkscape. When you spot a 'duo' you can merge the point.

The result of this .svg inside inkcut will be:

* trace to point A
* pen UP to point B
* trace to point C

with a zero displacement between A and B (they are identical) but a real PU PD

IN;PU12146,3849;PD11632,3217;PD11567,3187;PD11508,3176;PD11454,3179;PD11407,3191;PD11334,3226;PD11293,3245;PD11298,2523;PD11298,2479;PD11315,2443;PD11328,2432;PD11345,2429;PD11366,2436;PD11389,2454;PD12401,3555;PD13397,2230;PD13418,2223;PD13437,2223;PD13470,2239;PD13493,2262;PD13503,2279;PD13469,3256;PD13385,3203;PD13341,3183;PD13295,3170;PD13247,3169;PD13197,3182;PD13143,3214;PD13085,3266; ** **PD12620,3849;PU12620,3849;** ** PD13209,4541;PD13252,4576;PD13291,4594;PD13326,4600;PD13357,4596;PD13406,4577;PD13437,4564;PD13433,5269;PD13425,5297;PD13409,5304;PD13377,5293;PD12365,4144;PD11333,5447;PD11299,5458;PD11278,5450;PD11263,5420;PD11261,4462;PD11354,4511;PD11406,4528;PD11461,4535;PD11518,4532;PD11577,4514;PD11636,4480;PD11696,4427;PD12146,3849;

Inkcut should optionally remove those 0 zero moves.

### LINKED ISSUE #355: Inconsistent device filter order
When working on a filter for min line, and min  jump length I noticed that order in which filters get applied are inconsistent. Sometimes my new filter got applied before blade offset filter and sometimes after.

After a bit of debugging and code reading I found that the application order depends on the order in which the filters where enabled. This happens because when a filter gets enabled it gets placed at the end of list. https://github.com/inkcut/inkcut/blob/35985e524a97989f5b9e13edd2fe859fe4f34a1b/inkcut/device/dialogs.enaml#L207 

In theory refresh_filter uses ext.rank but it is currently ineffective due to the code I linked above. Ext.rank also becomes problematic to use when single DEVICE_FILTER_POINT `Extension`  declares multiple `DeviceFilter` because the single rank applies to all those filters. But it shouldn't be difficult to reorganize manifest so that it uses separate Extension for each filter each with it's own rank.

Ideas for different approaches to deal with this:
* stable order based on order in plugin.filters -> solves the issue for builtin filters, doesn't help if user installs a bunch of third party plugins although I doubt Inkcut will have so many third party plugins for it to be a problem any time soon
* adding user configurable priority attribute to filter config (with sane defaults for builtin filters) -> a bit uglier and increases amount of stuff that can be misconfigured by user either accidentally or as a result of incremental settings upgrade
* use  filter tab order instead of user visible number input -> a bit nicer than previous approach, looking at enaml docs it wasn't obvious whether it can be easily queried

### PR DESCRIPTION
**Pull request by karliss**
* New filter for removing certain kind of short lines and gaps
   - Min jump: in combination with shortest path mode helps removing excessive pen up/down movements #273
   - min edge: removes short edges, can be used to simplify excessive subdivision in situations where curves where  already converted to line segments by the program generating svg. Not needed if svg contains actual curves as in that case you can just adjust quality setting in inkcut to control subdivision.
   - min path: removes short paths (connected sequence of edges without lifting pen).  Unlike min-edge  this won't  reduce smoothness of curves so it can be set higher. Useful to cleaning up small dots when plotting or some of the small features you know won't cut well.
   - min shift: cleans up connecting edges that would otherwise cause useless circle movements when using blade offset filter

Without min shift this:   
image
causes this:
image
That not only slows down the movement but can also cause small defects in the place where blade was spinning back and forth.

* Repeat filter
   - repeats each line specified number of times
   - useful when trying to cut cardborard which is on the edge of your machine capabilities
   - most effective with closed shapes making  multiple continuous  loops instead of whole drawing may be preferable over  simply running the whole program multiple times. Repeat filter approach reduces amount of tool lifts and also  errors caused by position slightly shifting over duration of large drawings. 
   - probably shouldn't be used when drawing if you want maximum consistency of line thickness
* Make the filter order more predictable #355
* In blade offset filter replace bad cubic bezier based arc approximation with qt api which does the same with simpler code and produces more accurate results.
Before arc fixes:
before
After arc fixes:
after

Closes #355
Closes #273

### CODE REVIEW

**Comment by karliss** [comment]
I just realized that for open shapes repeat filter could potentially move back and forth thus further decreasing tool lifts and traveling. But that should be optional because for cutting too many 180 degree turns in one place can cause slight defects, more so when blade offset isn't perfect or cutting thicker materials.

So I will leave that as an idea for future improvements.

File: inkcut/device/filters/repeat.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+Distributed under the terms of the GPL v3 License.
+
+The full license is in the file LICENSE, distributed with this software.
+
+Created on Dec 22, 2022
+
+@author: karliss
+"""
+from PyQt5.QtCore import QPointF
```
Comment (frmdstryr) [inline inkcut/device/filters/repeat.py:12]: Please use enaml.QtCore so it supports different Qt versions/backends

**Comment by frmdstryr** [comment]
Please reformat code with black. Nice clean implementations!  Thanks for contributing! 

P.S. sorry for the delay, I stopped watching the project so only randomly check up on it! Feel free to mention me in the future.

**Comment by karliss** [comment]
Done the things you asked.

> P.S. sorry for the delay, I stopped watching the project so only randomly check up on it! Feel free to mention me in the future.

Looks like Github finally added more fine grained notification controls, allowing to subscribe only to pull requests  instead of having to choose between "all, mentions, none" .

**Comment by karliss** [comment]
Closing the PR. Looks like it was merged manually by skipping the github PR system.

**Pull request closed without merging**